File: backend/app/schemas/generation.py
from pydantic import BaseModel, Field


class ConstraintConfig(BaseModel):
    min: float | None = None
    max: float | None = None
    min_age: int | None = None
    max_age: int | None = None
    values: str | None = None
    weights: str | None = None
    right_digits: int | None = None
    format: str | None = None
    start: str | None = None
    end: str | None = None


class FieldDefinition(BaseModel):
    name: str
    generator: str
    type: str = "string"
    unique: bool = False
    formula: str | None = None
    null_probability: float | None = None
    constraint: ConstraintConfig | None = None
    condition: str | None = None


class SharedKeyConfig(BaseModel):
    source_dataset: str
    source_field: str


class GroupConfig(BaseModel):
    num_groups: int = Field(..., ge=1, le=10000)
    split_pct: float = Field(default=100, ge=1, le=100)
    parent_fields: list[FieldDefinition]
    child_fields: list[FieldDefinition]


class DatasetDefinition(BaseModel):
    name: str
    template: str | None = None
    rows: int = Field(default=100, ge=1, le=100000)
    fields: list[FieldDefinition] = Field(default_factory=list)
    shared_key: SharedKeyConfig | None = None
    group_config: GroupConfig | None = None


class GenerateRequest(BaseModel):
    datasets: list[DatasetDefinition] = Field(
        ..., min_length=1, max_length=4
    )
    homogeneity: int = Field(default=50, ge=1, le=100)
    seed: int | None = None


class DatasetResult(BaseModel):
    dataset_id: str
    name: str
    table_name: str
    row_count: int
    columns: list[str]


class GenerateResponse(BaseModel):
    run_id: int
    homogeneity: int
    seed: int | None
    datasets: list[DatasetResult]

File: frontend/src/types/generation.ts
import type { ConstraintDef as ConstraintConfig, FieldDef, TemplateSummary } from "./template";

export type { ConstraintConfig, FieldDef, TemplateSummary };

export interface GroupConfig {
  num_groups: number;
  split_pct: number;
  parent_fields: FieldDef[];
  child_fields: FieldDef[];
}

export interface DatasetDefinition {
  name: string;
  template?: string | null;
  rows: number;
  fields: FieldDef[];
  shared_key?: SharedKeyConfig | null;
  group_config?: GroupConfig | null;
}

export interface SharedKeyConfig {
  source_dataset: string;
  source_field: string;
}

export interface FieldBreakConfig {
  field_name: string;
  break_rate: number;
  break_style: "drift" | "different" | "null";
  drift_pct: number;
}

export interface GenerateRequest {
  datasets: DatasetDefinition[];
  homogeneity: number;
  seed?: number | null;
  overlap_ratio?: number;
  exact_fields?: string[];
  reconciliation_mode?: boolean;
  field_breaks?: FieldBreakConfig[];
}

export interface DatasetResult {
  dataset_id: string;
  name: string;
  table_name: string;
  row_count: number;
  columns: string[];
}

export interface GenerateResponse {
  run_id: number;
  homogeneity: number;
  seed: number | null;
  datasets: DatasetResult[];
  overlap_pool_size: number;
  exact_fields: string[];
  break_count?: number;
}
